### ISSUE TITLE
ttnn.matmul hangs with fuse_batch=False + batched weight + block-sharded output

### ISSUE THREAD

**Issue by rpavlovicTT** [tenstorrent/tt-metal#42572]
## Summary

`ttnn.matmul` with `MatmulMultiCoreReuseMultiCastProgramConfig(fuse_batch=False)` **hangs** (no error, no timeout) when:

1. Both inputs have a non-unit batch dim (`in1` batch != 1), so `fuse_batch=False` is the only legal choice — `fuse_batch=True` is rejected by the shape check at `matmul_device_operation.cpp:393`, AND
2. The output `memory_config` is **block-sharded on the compute grid**.

Switching the output memory_config to interleaved L1 makes the same call complete in milliseconds.

Expected: the matmul either completes successfully or is rejected at `validate()` time with a descriptive error.

## Environment

- Hardware: Wormhole B0, single device
- tt-metal: `v0.69.0-dev20260414-8-g90c914ef25`
- Python 3.12, Ubuntu 24.04

## Minimal reproducer

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)
try:
    in0_t = torch.randn((4, 32, 2880), dtype=torch.bfloat16)
    in1_t = torch.randn((4, 2880, 5760), dtype=torch.bfloat16)

    in0 = ttnn.from_torch(
        in0_t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
        device=device, memory_config=ttnn.L1_MEMORY_CONFIG,
    )
    in1 = ttnn.from_torch(
        in1_t, dtype=ttnn.bfloat8_b, layout=ttnn.TILE_LAYOUT,
        device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )

    program_config = ttnn.MatmulMultiCoreReuseMultiCastProgramConfig(
        compute_with_storage_grid_size=ttnn.CoreCoord(8, 4),
        in0_block_w=2,
        out_subblock_h=1, out_subblock_w=1,
        out_block_h=1,   out_block_w=23,
        per_core_M=1,    per_core_N=23,
        transpose_mcast=False,
        fuse_batch=False,
    )

    block_sharded_out = ttnn.MemoryConfig(
        memory_layout=ttnn.TensorMemoryLayout.BLOCK_SHARDED,
        buffer_type=ttnn.BufferType.L1,
        shard_spec=ttnn.ShardSpec(
            ttnn.CoreRangeSet(
                [ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(7, 3))]
            ),
            (1 * 32, 23 * 32),
            ttnn.ShardOrientation.ROW_MAJOR,
        ),
    )

    out = ttnn.matmul(
        in0, in1,
        program_config=program_config,
        dtype=ttnn.bfloat16,
        memory_config=block_sharded_out,    # replace with ttnn.L1_MEMORY_CONFIG -> completes
    )
    ttnn.synchronize_device(device)
    print("Matmul done.", out.shape)
finally:
    ttnn.close_device(device)
```
